include/pybind11/cast.h

```diff
@@ -208,13 +208,21 @@ class type_caster_generic {
             }
         }
 
-        /* Perform an implicit conversion */
+        /* Perform an implicit or a direct conversion */
         if (convert) {
             for (auto &converter : typeinfo->implicit_conversions) {
                 temp = object(converter(src.ptr(), typeinfo->type), false);
                 if (load(temp, false))
                     return true;
             }
+            auto& direct = get_internals().direct_conversions;
```

Comment:
Let's not call ``get_internals()`` over and over again in this function, we can cache it.